pkg/plugins/storage/storage/storage.go

```diff
@@ -179,6 +198,85 @@ func (s *StorageStorageService) Delete(bucket string, key string) error {
 	return nil
 }
 
+func (s *StorageStorageService) PreSignUrl(bucket string, key string, operation plugin.Operation, expiry uint32) (string, error) {
+	newErr := errors.ErrorsWithScope(
+		"StorageStorageService.PreSignUrl",
+		map[string]interface{}{
+			"bucket":    bucket,
+			"key":       key,
+			"operation": operation.String(),
+		},
+	)
+	if len(bucket) == 0 {
+		return "", newErr(
+			codes.InvalidArgument,
+			"provide non-blank bucket",
+			fmt.Errorf(""),
+		)
+	}
+	if len(key) == 0 {
+		return "", newErr(
+			codes.InvalidArgument,
+			"provide non-blank key",
+			fmt.Errorf(""),
+		)
+	}
+	if expiry > 604800 {
+		return "", newErr(
+			codes.InvalidArgument,
+			"expiry time can't be longer than 604800 seconds (7 days)",
+			fmt.Errorf("expiry provided: %d", expiry),
+		)
+	}
+	serviceAccount := utils.GetEnv("SERVICE_ACCOUNT_LOCATION", "")
+	if len(serviceAccount) == 0 {
+		return "", newErr(
+			codes.InvalidArgument,
+			"provide non-blank service account config path",
+			fmt.Errorf("path provided: %s", serviceAccount),
+		)
+	}
+	jsonKey, err := ioutil.ReadFile(serviceAccount)
+	if err != nil {
+		return "", newErr(
+			codes.Internal,
+			"error reading service account config file at given SERVICE_ACCOUNT_LOCATION",
+			err,
+		)
+	}
```

Comment:
I think the credentials can be retrieved using the same method we use to construct the client:

`google.FindDefaultCredentials(ctx, storage.ScopeReadWrite)`

Maybe store them privately on the StorageStorageService struct and reuse them here instead of looking in a custom location.

The returned value has a `credentials.JSON` property, which you can probably use with the remaining code.


Interestingly, other Google Cloud SDKs like Node don't ask for the `PrivateKey` or `GoogleAccessID` when signing URLs. Maybe double-check how they're implemented and we can copy that behaviour.

https://cloud.google.com/storage/docs/access-control/signing-urls-with-helpers#code-samples